add simple history provider

MatthewRBruce opened the pull request:
## What changes were proposed in this pull request?

This change is to add a History Provider that doesn't rely on scanning the entirely of the event-logs directory to display what histories are available.  Instead the SimpleFsHistoryProvider relies on the application ID provided in the URL to determine which event logs the user would like to view.

## How was this patch tested?

Still need to write unit tests - additional commit

MatthewRBruce committed:
add simple history provider

angelini commented:
As additional context:

- We've never used the index page provided by the Spark history server and instead generate direct links for all jobs.
- The preprocessing of logs often gets stuck at our scale, which blocks one of the best debug tools from the team.
- Just this past week when we tried to load a job history on the production server it failed because the cache had an invalid item. @MatthewRBruce's version worked great and was able to unblock the person debugging their job.
- We've had issues with log processing in the Spark history server since version 1.4.

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+        )
+        Some(attemptInfo)
+      }
+      case None => None
```
You can avoid this pattern:

```
match {
  case Some(...) => Some(do something)
  case None => None
}
```

by mapping over the option. `Option.map(f)` will apply `f` to the contents of the `Option` if they are set otherwise it will return `None`.

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+          Some(new FsApplicationHistoryInfo(attemptInfo.appId, attemptInfo.name, List(attemptInfo)))
+        case None =>
+          None
+      }
```
You can also use `Option.map` here

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+            case None => None
+          }
+        }
+        case None => None
```
More cases of `Option.map`

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+      a1: FsApplicationAttemptInfo,
+      a2: FsApplicationAttemptInfo): Boolean = {
+    a1.startTime >= a2.startTime
+  }
```
It looks like this is dead code.

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+    val a1 = i1.attempts.head
+    val a2 = i2.attempts.head
+    if (a1.endTime != a2.endTime) a1.endTime >= a2.endTime else a1.startTime >= a2.startTime
+  }
```
Also looks unused

angelini commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,308 @@
+      directory for existing event logs.  Please browse to
+      history/<span style="font-style:italic">applicationId</span>
+      to view the UI for a particular application
+    </p>
```
Inline HTML, oh you crazy Scala

MatthewRBruce committed:
Refator

- Switch some match case statements for maps
- Remove some dead code
- Put in object for SimpleFsHistoryProvider that contains some contants

MatthewRBruce committed:
missed a match case

andrewdryder commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,277 @@
+ */
+
+package org.apache.spark.deploy.history
+
```
A lot of the listed imports aren't actually used - can you remove them?

pfryerdavis replied:
Follow up to that, can you get rid of the commented out imports?

andrewdryder commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+  //Return an empty list iterator
+  override def getListing(): Iterator[FsApplicationHistoryInfo] = List().iterator
+  
+  def getApplicationAttemptInfo(appFileStatus: FileStatus): Option[org.apache.spark.deploy.history.FsApplicationAttemptInfo] = {
```
`Option[FsApplicationAttemptInfo]`

andrewdryder commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+    "; groups with admin permissions" + HISTORY_UI_ADMIN_ACLS_GROUPS.toString)
+
+  private val hadoopConf = SparkHadoopUtil.get.newConfiguration(conf)
+  private val fs = new Path(logDir).getFileSystem(hadoopConf)
```
This seems awkward - why not `private val fs = <org.apache.hadoop.fs.>FileSystem.get(hadoopConf)`?

pseudomuto commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+}
+
+private[history] object SimpleFsHistoryProvider {
+  val DEFAULT_LOG_DIR = "file:/tmp/spark-events"
```
Probably not a thing, but this doesn't require that weird triple slash dealy right? i.e. `file:///tmp/spark-events`

MatthewRBruce replied:
that's the same as in the FsHistoryProvider, so I guess it's safe

pseudomuto commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+        case e: Exception =>
+          logError(s"Exception encountered when attempting to load application log ${logDir}/${appId}", e)
+          None
+        case e: FileNotFoundException =>
```
Ordering isn't an issue here, right?

MatthewRBruce replied:
yeah... it probably is

pseudomuto commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+  //Return an empty list iterator
+  override def getListing(): Iterator[FsApplicationHistoryInfo] = List().iterator
+  
+  def getApplicationAttemptInfo(appFileStatus: FileStatus): Option[org.apache.spark.deploy.history.FsApplicationAttemptInfo] = {
```
/nit why the fully qualified name here? Name conflict or something?

MatthewRBruce replied:
yeah, I had problems with the import, but it seems fine without it now.

andrewdryder commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+     * the name of the file being compressed.
+     */
+    def zipFileToStream(file: Path, entryName: String, outputStream: ZipOutputStream): Unit = {
+      val fs = file.getFileSystem(hadoopConf)
```
do we need to get another FS?

MatthewRBruce replied:
This is part of the untouched code from FsHistoryProvider, but it looks like you make Path  object, then call get getFileSystem.  Seems kind of backwards, but that appears to be how it works.  This FS object will be specific to the application's event log's dir vs the whole event-logs logdir

andrewdryder replied:
Makes sense now - still looks D:

andrewdryder commented on core/src/main/scala/org/apache/spark/deploy/history/SimpleFsHistoryProvider.scala:
```diff
@@ -0,0 +1,273 @@
+     */
+    def zipFileToStream(file: Path, entryName: String, outputStream: ZipOutputStream): Unit = {
+      val fs = file.getFileSystem(hadoopConf)
+      val inputStream = fs.open(file, 1 * 1024 * 1024) // 1MB Buffer
```
`1 * 1024 * 1024` should probably be a constant

MatthewRBruce replied:
I can, this is part of the code that came from FsHistoryProvider

andrewdryder reviewed (changes requested):
Are the scalastyle rules provided by us or is it default for the spark project? Might be worth just checking the line-lengths and such (if the ultimate goal is to push this back into the Spark project)

pseudomuto reviewed (commented):
All I've got are nits

MatthewRBruce committed:
refactor

angelini commented:
Is this good to merge?

This would be super useful to have in production.

andrewdryder commented:
The changes requested are fairly superficial, this could probably be merged

MatthewRBruce commented:
I'll get a build of this into our spark history image and get a test deploy out in adhoc
